File: neatformcore/src/main/java/com/opensrp/neatformcore/views/controls/EditTextNFormView.java
package com.opensrp.neatformcore.views.controls;

import android.content.Context;
import android.support.v7.widget.AppCompatEditText;
import android.util.AttributeSet;

import com.opensrp.neatformcore.domain.model.NFormViewData;
import com.opensrp.neatformcore.domain.model.NFormViewOption;
import com.opensrp.neatformcore.domain.model.NFormViewProperty;
import com.opensrp.neatformcore.domain.view.DataPassListener;
import com.opensrp.neatformcore.domain.view.NFormView;
import com.opensrp.neatformcore.utils.Constants;
import com.opensrp.neatformcore.views.data.ViewDataHandler;

import static com.opensrp.neatformcore.form.json.NFormViewBuilder.makeView;

public class EditTextNFormView extends AppCompatEditText implements NFormView {

    private static String TAG = EditTextNFormView.class.getCanonicalName();

    private NFormViewProperty viewProperty;

    private DataPassListener dataPassListener;

    private NFormViewOption viewOption;

    public EditTextNFormView(Context context) {
        super(context);
    }

    public EditTextNFormView(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public EditTextNFormView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    @Override
    public NFormViewOption getViewOption() {
        return viewOption;
    }

    @Override
    public NFormView initView(NFormViewProperty viewProperty, ViewDataHandler viewDataHandler) {
        this.viewOption = new NFormViewOption(this);
        if (viewProperty != null) {
            viewOption.setName(viewProperty.getName());

            if (viewProperty.getViewAttributes() != null) {
                makeView(viewProperty.getViewAttributes(), this, Constants.ViewType.EDIT_TEXT);
            }
        }
        setOnDataPassListener(viewDataHandler);
        return this;
    }

    @Override
    public NFormViewData getViewData() {
        return new NFormViewData();
    }

    @Override
    public void setOnDataPassListener(DataPassListener dataPassListener) {
        if (this.dataPassListener == null) {
            this.dataPassListener = dataPassListener;
        }
    }

    @Override
    protected void onTextChanged(CharSequence text, int start, int lengthBefore, int lengthAfter) {
        super.onTextChanged(text, start, lengthBefore, lengthAfter);

        if (viewOption != null && this.dataPassListener != null) {
            this.viewOption.setValue(text);
            this.dataPassListener.onPassData(viewOption);
        }
    }
}

File: neatformcore/src/main/java/com/nerdstone/neatformcore/views/widgets/EditTextNFormView.kt
package com.nerdstone.neatformcore.views.widgets

import android.content.Context
import android.support.v7.widget.AppCompatEditText
import android.util.AttributeSet
import android.view.View
import com.nerdstone.neatformcore.domain.data.DataActionListener
import com.nerdstone.neatformcore.domain.model.NFormViewData
import com.nerdstone.neatformcore.domain.model.NFormViewDetails
import com.nerdstone.neatformcore.domain.model.NFormViewProperty
import com.nerdstone.neatformcore.domain.view.NFormView
import com.nerdstone.neatformcore.domain.view.RootView
import com.nerdstone.neatformcore.domain.view.RulesHandler
import com.nerdstone.neatformcore.utils.ViewUtils
import com.nerdstone.neatformcore.views.builders.EditTextViewBuilder
import com.nerdstone.neatformcore.views.handlers.ViewDispatcher

class EditTextNFormView : AppCompatEditText, NFormView {

    override lateinit var viewProperties: NFormViewProperty
    override var dataActionListener: DataActionListener? = null
    override val viewBuilder: EditTextViewBuilder = EditTextViewBuilder(this)
    override var viewDetails: NFormViewDetails = NFormViewDetails(this)
    override val viewData get() = NFormViewData()
    override val nFormRootView get() = this.parent as RootView

    constructor(context: Context) : super(context)

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs)

    override fun initView(viewProperty: NFormViewProperty, viewDispatcher: ViewDispatcher)
            : NFormView {
        ViewUtils.setupView(this, viewProperty, viewDispatcher)
        return this
    }

    override fun mapViewIdToName(rulesHandler: RulesHandler) {
        id = View.generateViewId()
        rulesHandler.viewIdsMap[this.viewProperties.name] = id
    }

    override fun setOnDataPassListener(dataActionListener: DataActionListener) {
        if (this.dataActionListener == null) this.dataActionListener = dataActionListener
    }

    override fun onTextChanged(
        text: CharSequence, start: Int, lengthBefore: Int,
        lengthAfter: Int
    ) {
        super.onTextChanged(text, start, lengthBefore, lengthAfter)
        if (text.isNotEmpty()) {
            dataActionListener?.also {
                this.viewDetails.value = text.toString()
                it.onPassData(viewDetails)
            }
        }
    }
}
